qstash/quickstarts/vercel-nextjs.mdx

```diff
@@ -1,204 +1,419 @@
+    const data = await request.json()
+
+    for (let i = 0; i < 10; i++) {
+      await fetch('https://firstqstashmessage.requestcatcher.com/test', {
+        method: 'POST',
+        body: JSON.stringify(data),
+        headers: { 'Content-Type': 'application/json' }
+      })
+      await new Promise(resolve => setTimeout(resolve, 500))
+    }
+
+
+    return Response.json({ success: true })
+  }
+  ```
+
+  Now let's update our `startBackgroundJob` function to use our new endpoint.
+
+  There's 1 problem: our endpoint is not public. We need to make it public so that QStash can call it.
+  We have 2 options:
+
+  1. Deploy our application to a platform like Vercel and use the public URL.
+  2. Create a [local tunnel](/qstash/howto/local-tunnel) to test the endpoint locally.
+
+  For the purpose, of this tutorial, I'll deploy the application to Vercel, but 
+  feel free to use a local tunnel if you prefer.
+
+  <Accordion title="Deploying to Vercel">
+    There are many ways to [deploy to Vercel](https://vercel.com/docs/deployments/overview), but
+    I'm going to use the Vercel CLI.
+
+    ```bash
+    npm install -g vercel
+    ```
+
+    ```bash
+    vercel
+    ```
+
+    Once deployed, you can find the public URL in the Vercel dashboard.
+  </Accordion>
+
+  Now that we have a public URL, we can update the URL.
+
+  ```ts src/app/actions.ts
+  "use server"
+  import { Client } from "@upstash/qstash";
+
+  const qstashClient = new Client({
+    token: process.env.QSTASH_TOKEN!,
+  });
+
+  export async function startBackgroundJob() {
+    await qstashClient.publishJSON({
+      // Replace with your public URL
+      "url": "https://qstash-bg-job.vercel.app/api/long-task",
+      body: {
+        "hello": "world"
+      }
+    })
+  }
+  ```
+
+  And voila! You've created a Next.js app that calls a long-running background job using QStash.
+  </Step>
+
+  <Step title="Error catching and security">
+  QStash is a great way to handle background jobs, but it's important to remember that it's a public
+  API. This means that anyone can call your endpoint. Make sure to add security measures to your endpoint
+  to ensure that QStash is the sender of the request.
+
+  Luckily, our SDK provides a way to verify the sender of the request. Make sure to get your signing keys
+  from the QStash console and add them to your environment variables.
+
+
+  ```ts src/app/api/long-task/route.ts
+  import { Receiver } from "@upstash/qstash/";
+
+  const receiver = new Receiver({
+    currentSigningKey: process.env.QSTASH_CURRENT_SIGNING_KEY!,
+    nextSigningKey: process.env.QSTASH_NEXT_SIGNING_KEY!
+  })
+
+  export async function POST(request: Request) {
+    const signature = request.headers.get("Upstash-Signature");
+    if (!signature) {
+      return Response.json({ error: "No signature" }, { status: 400 });
+    }
+    const data = await request.json();
+
+    // This will throw an error if the signature is invalid
+    receiver.verify({
```

Comment:
We can actually do the verification like this
```
import { NextRequest, NextResponse } from "next/server";
import { verifySignatureAppRouter } from "@upstash/qstash/dist/nextjs";

async function handler(_req: NextRequest) {
    await new Promise((r) => setTimeout(r, 1000));

    console.log("Success");
    return NextResponse.json({ name: "John Doe Serverless" });
}

export const POST = verifySignatureAppRouter(handler);
```

Reply:
Both of these guides are so well written, thanks. We could maybe change the page title to something more simpler. I think the content is simple enough to be a quick start, but the title is a bit specific for a quick start at the top of the list. What do you think?

Also, are we going to keep the `Python` quick start?

Reply:
> We can actually do the verification like this
> 
> ```
> import { NextRequest, NextResponse } from "next/server";
> import { verifySignatureAppRouter } from "@upstash/qstash/dist/nextjs";
> 
> async function handler(_req: NextRequest) {
>     await new Promise((r) => setTimeout(r, 1000));
> 
>     console.log("Success");
>     return NextResponse.json({ name: "John Doe Serverless" });
> }
> 
> export const POST = verifySignatureAppRouter(handler);
> ```

Good point, I'll fix.


> Both of these guides are so well written, thanks. We could maybe change the page title to something more simpler. I think the content is simple enough to be a quick start, but the title is a bit specific for a quick start at the top of the list. What do you think?
> 
> Also, are we going to keep the `Python` quick start?

Sounds good, I'll change it to something more vague like `Next.js` and `Python`. I can replace the existing python one, since that one just covers how to implement verification which isn't needed with the SDK.

Reply:
> Both of these guides are so well written, thanks. We could maybe change the page title to something more simpler. I think the content is simple enough to be a quick start, but the title is a bit specific for a quick start at the top of the list. What do you think?
> 
> Also, are we going to keep the `Python` quick start?

After looking again, I generalized the title to be `Next.js` and I think it's best to keep the other Python one as well since it shows how to create it on lambda. So I renamed it to `Python on AWS` and `Python on Vercel` respectively.